### PR TITLE
[Kernel] Add FlashInfer TRTLLM MXFP8 linear backend

### PR DESCRIPTION
**Pull request by seonjinn**
## Purpose

FlashInfer exposes a TensorRT-LLM backend for dense MXFP8 GEMM, but vLLM's
`flashinfer_trtllm` linear selector currently covers NVFP4 only. This change
adds `FlashInferTrtllmMxfp8LinearKernel`, which can be selected with:

```text
--linear-backend flashinfer_trtllm
```

The kernel prepares TensorRT-LLM weight and scale layouts once after weight
loading. At runtime it quantizes BF16 activations with the 8x4 scale layout,
calls FlashInfer's TensorRT-LLM MXFP8 GEMM, and slices the physically padded
output back to its logical width before adding bias.

The kernel is the last CUDA entry in the MXFP8 priority list, so existing
kernels retain their normal priority. Explicit `flashinfer_trtllm` selection
filters the list to this kernel. This PR does not add tactic hints, MoE kernel
changes, or the high-M 128x4 activation-scale path.

No matching open PR was found with the searches `MXFP8 TRTLLM linear` and
`mm_mxfp8 trtllm`. The closest merged implementations are the NVFP4 TRTLLM
linear backend (#39129) and the MXFP8 CuTeDSL linear backend (#46393). Open PR
#52016 adds the independent B12X CuTeDSL package for SM120/SM121 and does not
expose FlashInfer's TensorRT-LLM MXFP8 runner.

## Test Result

```text
.venv/bin/python -m pytest --confcutdir=tests/kernels/quantization \
  tests/kernels/quantization/test_mxfp8_trtllm_linear.py -q

1 passed
```

```text
.venv/bin/pre-commit run --files \
  docs/features/quantization/modelopt.md \
  tests/kernels/quantization/test_flashinfer_mxfp8_trtllm.py \
  tests/kernels/quantization/test_mxfp8_trtllm_linear.py \
  vllm/model_executor/kernels/linear/__init__.py \
  vllm/model_executor/kernels/linear/mxfp8/flashinfer.py

All hooks passed
```

The GB200 test passed five real-GPU cases: three numerical shapes with M=1,
7, and 128, custom-op checks, and CUDA Graph capture and replay. The N=130
cases cover physical output padding and slicing. The run used commit
`37d19a7edd7559c3f8da24703d53feea198f96b4` and FlashInfer 0.6.16.post3,
matching the vLLM dependency pin. Model-level accuracy and same-version
performance results are not yet available, so this PR makes no accuracy or
performance claim.

### CODE REVIEW

**Comment by seonjinn** [comment]
@claude review

File: docs/features/quantization/modelopt.md [hunk 1]
```diff
@@ -33,6 +33,12 @@ following `quantization.quant_algo` values:
     [Engine Arguments](../../configuration/engine_args.md) page and shown by
     `vllm serve --help=KernelConfig`.
 
+!!! note
+    For MXFP8 checkpoints on SM100, SM103, and SM107 GPUs, BF16 models can
```
Comment (mgoin) [inline docs/features/quantization/modelopt.md]: Do you mean "MXFP8 models"?

File: vllm/model_executor/kernels/linear/__init__.py [hunk 2]
```diff
@@ -840,7 +842,14 @@ def init_mxfp8_linear_kernel() -> Mxfp8LinearKernel:
...
     platform = current_platform._enum
-    possible = list(_POSSIBLE_MXFP8_KERNELS.get(platform, []))
+    possible = [
+        kernel
+        for kernel in _POSSIBLE_MXFP8_KERNELS.get(platform, [])
+        if kernel is not FlashInferTrtllmMxfp8LinearKernel
+    ]
+
+    if _get_linear_backend() == "flashinfer_trtllm":
+        possible.append(FlashInferTrtllmMxfp8LinearKernel)
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/__init__.py]: Maybe you could add it to the bottom of _POSSIBLE_MXFP8_KERNELS if you don't want this chosen by default? I'd prefer to not have a special case

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py [hunk 3]
```diff
@@ -177,3 +177,111 @@ def apply_weights(
...
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        if x.dtype != torch.bfloat16:
+            raise ValueError(
+                "FlashInfer TRTLLM MXFP8 requires bfloat16 output, "
+                f"got input dtype {x.dtype}."
+            )
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp8/flashinfer.py]: nit: assert

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py [hunk 4]
```diff
@@ -177,3 +177,111 @@ def apply_weights(
...
+            input_mxfp8,
+            weight.t(),
+            input_scale,
+            weight_scale,
+            out_dtype=x.dtype,
+            backend="trtllm",
+            use_8x4_sf_layout=True,
+        )
+        if output.shape[-1] != output_size:
+            output = output[:, :output_size].contiguous()
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp8/flashinfer.py:273]: Are we missing padding the input to match the weight?

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py [hunk 5]
```diff
@@ -177,3 +177,111 @@ def apply_weights(
...
+            return False, "requires CUDA"
+        if compute_capability is None:
+            supported_capability = any(
+                current_platform.is_device_capability(capability)
+                for capability in (100, 103, 107)
+            )
+        else:
+            supported_capability = compute_capability in (100, 103, 107)
+        if not supported_capability:
+            return False, "requires sm_100/sm_103/sm_107 (Blackwell)"
```
Comment (mgoin) [inline vllm/model_executor/kernels/linear/mxfp8/flashinfer.py]: nit: can just do `current_platform.is_device_capability_family(100)`

File: tests/kernels/quantization/test_mxfp8_trtllm_linear.py
Comment (mgoin) [inline tests/kernels/quantization/test_mxfp8_trtllm_linear.py:1]: I think a lot of these mocked tests are overkill

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py (hunk 5 above)
Comment (seonjinn) [reply vllm/model_executor/kernels/linear/mxfp8/flashinfer.py]: Oh, okay.

File: tests/kernels/quantization/test_mxfp8_trtllm_linear.py
Comment (seonjinn) [reply tests/kernels/quantization/test_mxfp8_trtllm_linear.py:1]: I removed this part

Comment (seonjinn) [reply tests/kernels/quantization/test_mxfp8_trtllm_linear.py:1]: Reduced the mocked CPU suite from 252 lines to one backend-selection test. Numerics, padding, custom-op checks, and CUDA Graph coverage remain in the real GPU tests.

File: vllm/model_executor/kernels/linear/__init__.py (hunk 2 above)
Comment (seonjinn) [reply vllm/model_executor/kernels/linear/__init__.py]: Done. The kernel is now the last CUDA MXFP8 candidate, and the special-case selection logic is removed.

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py (hunk 4 above)
Comment (seonjinn) [reply vllm/model_executor/kernels/linear/mxfp8/flashinfer.py:273]: No input padding is needed. Only N is padded, so the operands remain A[M, K] and B[K, padded_N]; the result is sliced back to N. The 8x4 quantizer handles M-side scale-layout padding internally.

File: docs/features/quantization/modelopt.md (hunk 1 above)
Comment (seonjinn) [reply docs/features/quantization/modelopt.md]: Updated to "models quantized to MXFP8 with BF16 activations" to distinguish the weight format from the activation dtype.

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py (hunk 3 above)
Comment (seonjinn) [reply vllm/model_executor/kernels/linear/mxfp8/flashinfer.py]: Done. Replaced the BF16 dtype ValueError with an assertion.

File: vllm/model_executor/kernels/linear/mxfp8/flashinfer.py (hunk 5 above)
Comment (seonjinn) [reply vllm/model_executor/kernels/linear/mxfp8/flashinfer.py]: Done. The kernel and its GPU test now use the SM100-family capability helper.

**Review by mgoin — APPROVED** [review]
LGTM thanks!

**Comment by seonjinn** [comment]
@mgoin For amd-mi300-v1-spec-decode test case, does it take longer time usually?
